File: frontend/web-app/src/types/fund.ts
export interface Fund {
  fund_id: string;
  fund_name: string;
  fund_type: 'private_equity' | 'venture_capital' | 'hedge_fund' | 'real_estate';
  fund_status: 'active' | 'closed' | 'liquidated';
  inception_date: string;
  target_size: number;
  committed_capital: number;
  called_capital: number;
  paid_in_capital: number;
  nav: number;
  irr: number;
  moic: number;
  management_fee_rate: number;
  carry_rate: number;
  description?: string;
  created_at: string;
  updated_at: string;
}

export interface FundPerformance {
  fund_id: string;
  as_of_date: string;
  nav: number;
  irr: number;
  moic: number;
  tvpi: number;
  dpi: number;
  rvpi: number;
  called_capital: number;
  distributed_capital: number;
  remaining_value: number;
}

export interface CapitalCall {
  call_id: string;
  fund_id: string;
  call_date: string;
  due_date: string;
  total_amount: number;
  call_percentage: number;
  status: 'pending' | 'completed' | 'overdue';
  investor_calls: InvestorCapitalCall[];
}

export interface InvestorCapitalCall {
  investor_id: string;
  call_amount: number;
  paid_amount: number;
  payment_date?: string;
  status: 'pending' | 'paid' | 'overdue';
}

export interface Distribution {
  distribution_id: string;
  fund_id: string;
  distribution_date: string;
  total_amount: number;
  distribution_type: 'return_of_capital' | 'capital_gains' | 'dividend';
  status: 'pending' | 'completed';
  investor_distributions: InvestorDistribution[];
}

export interface InvestorDistribution {
  investor_id: string;
  distribution_amount: number;
  tax_withholding: number;
  net_amount: number;
  payment_date?: string;
  status: 'pending' | 'paid';
}

export interface CreateFundRequest {
  fund_name: string;
  fund_type: Fund['fund_type'];
  inception_date: string;
  target_size: number;
  management_fee_rate: number;
  carry_rate: number;
  description?: string;
}

export interface UpdateFundRequest extends Partial<CreateFundRequest> {
  fund_id: string;
}

File: backend/api-gateway/src/models/fund.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field
from enum import Enum


class FundType(str, Enum):
    PRIVATE_EQUITY = "private_equity"
    VENTURE_CAPITAL = "venture_capital"
    HEDGE_FUND = "hedge_fund"
    REAL_ESTATE = "real_estate"


class FundStatus(str, Enum):
    SETUP = "setup"
    ACTIVE = "active"
    CLOSED = "closed"
    LIQUIDATED = "liquidated"


class Fund(BaseModel):
    fund_id: str
    fund_name: str
    fund_type: FundType
    fund_status: FundStatus
    inception_date: datetime
    target_size: float
    committed_capital: float
    called_capital: float
    paid_in_capital: float
    nav: float
    irr: float
    moic: float
    management_fee_rate: float
    carry_rate: float
    description: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class FundPerformance(BaseModel):
    fund_id: str
    as_of_date: datetime
    nav: float
    irr: float
    moic: float
    tvpi: float
    dpi: float
    rvpi: float
    called_capital: float
    distributed_capital: float
    remaining_value: float


class InvestorCapitalCall(BaseModel):
    investor_id: str
    call_amount: float
    paid_amount: float
    payment_date: Optional[datetime] = None
    status: str


class CapitalCall(BaseModel):
    call_id: str
    fund_id: str
    call_date: datetime
    due_date: datetime
    total_amount: float
    call_percentage: float
    status: str
    investor_calls: List[InvestorCapitalCall] = []


class InvestorDistribution(BaseModel):
    investor_id: str
    distribution_amount: float
    tax_withholding: float
    net_amount: float
    payment_date: Optional[datetime] = None
    status: str


class Distribution(BaseModel):
    distribution_id: str
    fund_id: str
    distribution_date: datetime
    total_amount: float
    distribution_type: str
    status: str
    investor_distributions: List[InvestorDistribution] = []


class CreateFundRequest(BaseModel):
    fund_name: str = Field(..., min_length=2, max_length=100)
    fund_type: FundType
    inception_date: datetime
    target_size: float = Field(..., gt=0)
    management_fee_rate: float = Field(..., ge=0, le=1)
    carry_rate: float = Field(..., ge=0, le=1)
    description: Optional[str] = None


class UpdateFundRequest(BaseModel):
    fund_id: str
    fund_name: Optional[str] = Field(None, min_length=2, max_length=100)
    fund_type: Optional[FundType] = None
    inception_date: Optional[datetime] = None
    target_size: Optional[float] = Field(None, gt=0)
    management_fee_rate: Optional[float] = Field(None, ge=0, le=1)
    carry_rate: Optional[float] = Field(None, ge=0, le=1)
    description: Optional[str] = None